### ISSUE TITLE
Performance issue with `nvim_get_mode`

### ISSUE THREAD

**Issue by pierreganty** [neovim/neovim#19352]
### Neovim version (nvim -v)

v0.8.0-dev+604-gd8360e903

### Vim (not Nvim) behaves the same?

I did not check since vim does not have `pynvim`

### Operating system/version

macOS 12.4

### Terminal name/version

Terminal.app

### $TERM environment variable

xterm-256color

### Installation

build from repo

### How to reproduce the issue

I am running the following Python script (pynvim) to feed Neovim with three inputs (`i<C-o><C-g>`) and query the mode after each input. The script reports the times for each input. 
```python

import pynvim
from time import perf_counter as pc

n = pynvim.attach(
    "child", argv=['./nvim', '--clean', '--embed', '--headless', 'filename'])

print("Starting the sequence")

before = pc()
n.feedkeys(n.replace_termcodes("i"), 't')
assert (n.api.get_mode() == dict(
    {'mode': 'i', 'blocking': False})), "not in Insert mode or is blocked!"
after = pc()
print("i:", after - before)

before = pc()
n.feedkeys(n.replace_termcodes("<C-o>"), 't')
assert (n.api.get_mode() == dict(
    {'mode': 'niI', 'blocking': False})), "not in Insert Normal mode or is blocked!"
after = pc()
print("CTRL-O:", after - before)

before = pc()
n.feedkeys(n.replace_termcodes("<C-g>"), 't')
assert (n.api.get_mode() == dict(
    {'mode': 'i', 'blocking': False})), "not in Insert mode or is blocked!"
after = pc()
print("CTRL-G:", after - before)

print("End Of Sequence")
n.close()
```

### Expected behavior

The documentation says `nvim_get_mode()` is supposed to be served immediately.
```
Non-deferred ({fast}) functions such as |nvim_get_mode()| and
|nvim_input()| are served immediately (i.e. without waiting in the input
queue).
```

### Actual behavior

Yet I get the following output from the above script:
```
python3 -u ~/vim-model/trace.py
Starting the sequence
i: 1.0152065000000001
CTRL-O: 0.0016440410000000405
CTRL-G: 3.005534583
End Of Sequence
```
So `<C-G>` takes order of magnitude more time than the other two inputs which seems to go against what is stated in the documentation.

**Comment by zeertzjq** [comment]
This delay is caused by `'showmode'`. If you add a
```python
n.api.set_option('showmode', False)
```
before starting the sequence, this delay is gone.

**Comment by pierreganty** [comment]
Thank you. Do you also have a workaround for the similar case where instead of `<C-g>` you run a search for a pattern that is nowhere to be found?
```
python3 -u ~/vim-model/trace.py
Starting the sequence
i: 1.0151357079999999
CTRL-O: 0.0012680410000001086
/$blabla<CR>: 4.012448958
End Of Sequence
```

**Comment by pierreganty** [comment]
Btw you can see the issue for yourself even without the python script. Type `i` then `<C-O>` then `<C-G>`: it takes about 3 seconds until the `INSERT` mode appears at the end of the sequence.

**Comment by justinmk** [comment]
AFAIK, ctrl-g is special for some historical reason, and we don't yet process events during ctrl-g.

**Comment by zeertzjq** [comment]
No, this delay is hard-coded in `normal.c`:
<https://github.com/neovim/neovim/blob/778541067a2f0ec3ef4bc1b07def646323e852a9/src/nvim/normal.c#L642-L645=>

**Comment by zeertzjq** [comment]
I mean this workaround doesn't fully work for the error message case you mentioned.

**Comment by pierreganty** [comment]
Would that make sense to skip calls to os_delays for embedders _not_ using the UI protocol, that is when `nvim` is invoked with `--embed --headless`?
